### PR TITLE
Add atomic increment to FifoControllers

### PR DESCRIPTION
**Pull request by tguilbert-google**
This CL adds an increment counter methods that allow for atomically
incrementing the read/write indexes, rather than geting the current
index, adding to it and setting it.

Also, getFullFramesAvailable() and getEmptyFramesAvailable() now clamp
negative values to 0.

### CODE REVIEW

File: src/fifo/FifoController.h [hunk 1]
```diff
@@ -32,19 +32,24 @@ class FifoController : public FifoControllerBase
     FifoController(uint32_t bufferSize, uint32_t threshold);
     virtual ~FifoController() = default;
 
-    // TODO review use atomics or memory barriers
```
Comment (tguilbert-google) [inline src/fifo/FifoController.h]: From what I could tell, the current use of memory barriers is adequate to keep mWriteCounter and mReadCounter consistent. However, I don't know if there could be problems in FifoControllerBase::getFullFramesAvailable(), when we access both variables. I can't tell if there is a scenario where two threads would get a different number of frames, that could be prevented by a use of std::memory_order_seq_cst instead. From here (https://en.cppreference.com/w/cpp/atomic/memory_order#Sequentially-consistent_ordering) :
"Sequential ordering may be necessary for multiple producer-multiple consumer situations where all consumers must observe the actions of all producers occurring in the same order."

How many producers/consumers can FifoBuffer have?

Comment (tguilbert-google) [reply src/fifo/FifoController.h]: Useful video:
https://youtu.be/ZQFzMfHIxng?t=3117

Comment (philburk) [reply src/fifo/FifoController.h]: This FIFO has only one reader and one writer.

File: src/fifo/FifoBuffer.cpp [hunk 2]
```diff
@@ -85,7 +85,7 @@ int32_t FifoBuffer::convertFramesToBytes(int32_t frames) {
 }
 
 int32_t FifoBuffer::read(void *buffer, int32_t numFrames) {
-    int32_t framesAvailable = mFifo->getFullFramesAvailable();
+    uint32_t framesAvailable = mFifo->getFullFramesAvailable();
```
Comment (philburk) [inline src/fifo/FifoBuffer.cpp]: Please do not convert these to unsigned. See comments on https://github.com/google/oboe/pull/573

File: src/fifo/FifoControllerBase.cpp [hunk 3]
```diff
@@ -30,33 +30,34 @@ FifoControllerBase::FifoControllerBase(uint32_t totalFrames, uint32_t threshold)
 {
 }
 
-int32_t FifoControllerBase::getFullFramesAvailable() const {
-    return static_cast<int32_t>(getWriteCounter() - getReadCounter());
+uint32_t FifoControllerBase::getFullFramesAvailable() const {
```
Comment (philburk) [inline src/fifo/FifoControllerBase.cpp]: This value can go negative if there is an underflow! Please do not convert it to signed.

**Pull request merged**